File: ReSharper.FSharp/src/Daemon.FSharp/src/Stages/TypeCheckErrorsStage.cs
using JetBrains.Annotations;
using JetBrains.Application.Settings;
using JetBrains.ReSharper.Feature.Services.Daemon;
using JetBrains.ReSharper.Psi;
using JetBrains.ReSharper.Psi.FSharp.Tree;
using JetBrains.Util;
using Microsoft.FSharp.Compiler;

namespace JetBrains.ReSharper.Daemon.FSharp.Stages
{
  [DaemonStage(StagesBefore = new[] {typeof(SetResolvedSymbolsStage)},
    StagesAfter = new[] {typeof(IdentifiersHighlightStage)})]
  public class TypeCheckErrorsStage : FSharpDaemonStageBase
  {
    public class TypeCheckErrorsStageProcess : ErrorsStageProcess
    {
      public TypeCheckErrorsStageProcess([NotNull] IDaemonProcess process, [NotNull] FSharpErrorInfo[] errors)
        : base(process, errors)
      {
      }
    }

    protected override IDaemonStageProcess CreateProcess(IFSharpFile fsFile, IDaemonProcess process)
    {
      var errors = fsFile.CheckResults?.Errors ?? EmptyArray<FSharpErrorInfo>.Instance;
      return new TypeCheckErrorsStageProcess(process, errors);
    }

    public override ErrorStripeRequest NeedsErrorStripe(IPsiSourceFile sourceFile,
      IContextBoundSettingsStore settingsStore)
    {
      return ErrorStripeRequest.STRIPE_AND_ERRORS;
    }
  }
}

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/TypeCheckErrorsStage.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Daemon.Stages

open FSharp.Compiler.Diagnostics
open JetBrains.Application.Parts
open JetBrains.Diagnostics
open JetBrains.ReSharper.Feature.Services.Daemon
open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Stages
open JetBrains.Util

[<DaemonStage(Instantiation.DemandAnyThreadSafe, StagesBefore = [| typeof<SyntaxErrorsStage> |], StagesAfter = [| typeof<HighlightIdentifiersStage> |])>]
type TypeCheckErrorsStage(logger: ILogger) =
    inherit FSharpDaemonStageBase()

    override x.CreateStageProcess(fsFile, _, daemonProcess, _) =
        TypeCheckErrorsStageProcess(fsFile, daemonProcess, logger) :> _


and TypeCheckErrorsStageProcess(fsFile, daemonProcess, logger: ILogger) =
    inherit FcsErrorsStageProcessBase(fsFile, daemonProcess)

    let [<Literal>] opName = "TypeCheckErrorsStageProcess"

    override x.ShouldAddDiagnostic(error, range) =
        base.ShouldAddDiagnostic(error, range) &&

        match error.Subcategory with
        | "parse" ->
            // Directive syntax errors are only reported in type check results, so we allow these syntax errors
            isDirectiveSyntaxError error.ErrorNumber
        | _ -> true

    override x.Execute(committer) =
        match fsFile.GetParseAndCheckResults(false, opName) with
        | None -> ()
        | Some results ->

        let projectWarnings, fileErrors =
            results.CheckResults.Diagnostics
            |> Array.partition (fun e ->
                e.StartLine = 0 && e.EndLine = 0 && e.Severity = FSharpDiagnosticSeverity.Warning)

        if not (Array.isEmpty projectWarnings) then
            // https://github.com/Microsoft/visualfsharp/issues/4030
            let errors = projectWarnings |> Array.fold (fun result info -> result + "\n" + info.ToString()) ""
            logger.LogMessage(LoggingLevel.WARN, "Project warnings during file typeCheck:" + errors)

        x.Execute(fileErrors, committer)
